### PR TITLE
Aggregator

### PR DESCRIPTION
**Pull request by lukastanisic99**
First formal code review.

### CODE REVIEW

File: API/API.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.ServiceFabric.Services.Communication.AspNetCore;
+using Microsoft.ServiceFabric.Services.Communication.Runtime;
+using Microsoft.ServiceFabric.Services.Runtime;
+using Microsoft.ServiceFabric.Data;
+
+namespace API
```
Comment (GitTorre) [inline API/API.cs:15]: We should come up with a more descriptive name for this namespace. Naming in general will need some work, but this is minor.

File: Aggregator/Aggregator.cs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+                                s = null;
+                                data = null;
+                            }
+                        } 
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                var x = e;
```
Comment (GitTorre) [inline Aggregator/Aggregator.cs:98]: :-) (debugging...)

File: Aggregator/Aggregator.cs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        public async Task DeleteAllSnapshotsRemote()
+        {
+            using (var tx = this.StateManager.CreateTransaction())
+            {
+                try
+                {
+                    await (await this.StateManager.GetOrAddAsync<IReliableQueue<byte[]>>(Snapshot.queueName)).ClearAsync();
+                    await tx.CommitAsync();
+                }
+                catch(Exception e)
```
Comment (GitTorre) [inline Aggregator/Aggregator.cs:117]: Exception handling in general needs some work. This is a prototype, I understand that. Still, we should have some error handling in this codebase.

File: API/API.cs (hunk 1 above)
Comment (lukastanisic99) [reply API/API.cs:15]: Yes I agree. The API isn't really an API. it just calls the Aggregator API to fetch the data and displays it as plain text on a web page. Long term this service isn't needed and data can be ported to SFX or any other place.

File: Aggregator/Aggregator.cs (hunk 2 above)
Comment (lukastanisic99) [reply Aggregator/Aggregator.cs:98]: Yes this catch can be removed the catch is never triggered

File: Aggregator/Aggregator.cs (hunk 3 above)
Comment (lukastanisic99) [reply Aggregator/Aggregator.cs:117]: This function isn't used, the ClearAsync method isn't implemented for the reliable queue. Workaround is to dequeue elements one by one as used in pruning. I left this here if the method gets implemented in the future so it's just a simple call to clear all snapshots.

Comment (lukastanisic99) [reply Aggregator/Aggregator.cs:117]: But yes exception handling does need some work, but it's never triggered for now.

File: Aggregator/Collectors/CollectorBase.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                }
+            }
+        }
+
+        private async Task Batching(double intervalMiliseconds)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                T data =await CollectData();
+                lock (lockObj)
```
[the remark this answers is not in the document]
Comment (lukastanisic99) [reply Aggregator/Collectors/CollectorBase.cs:107]: The list is accessed by 2 threads. Line 54 starts a thread and the main thread also accesses it at line 67.

[the remark this answers is not in the document]
Comment (GitTorre) [reply Aggregator/Collectors/CollectorBase.cs:107]: Concurrently?

[the remark this answers is not in the document]
Comment (lukastanisic99) [reply Aggregator/Collectors/CollectorBase.cs:107]: Yes. It's an ever ongoing producer and consumer problem.

**Pull request closed without merging**